File: algorithms/284-peeking-iterator/peeking-iterator.py
class Iterator:
    def __init__(self, nums):
        """
        Initializes an iterator object to the beginning of a list.
        :type nums: List[int]
        """

    def hasNext(self):
        """
        Returns true if the iteration has more elements.
        :rtype: bool
        """

    def next(self):
        """
        Returns the next element in the iteration.
        :rtype: int
        """


class PeekingIterator:
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self._iter = iterator
        self._next = iterator.next()
        self._has_next = iterator.hasNext()

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        return self._next

    def next(self):
        """
        :rtype: int
        """
        ret = self._next
        self._has_next = self._iter.hasNext()
        if self._has_next:
            self._next = self._iter.next()
        else:
            self._next = 0
        return ret

    def hasNext(self):
        """
        :rtype: bool
        """
        return self._has_next

File: algorithms/284-peeking-iterator/peeking-iterator.go
package problem284

type Iterator struct{}

func (it *Iterator) hasNext() bool {
	// Returns true if the iteration has more elements.
	return false
}

func (it *Iterator) next() int {
	// Returns the next element in the iteration.
	return 0
}

type PeekingIterator struct {
	iter        *Iterator
	hashNext    bool
	nextElement int
}

// Constructor 构造 PeekingIterator 迭代器
func Constructor(iter *Iterator) *PeekingIterator {
	return &PeekingIterator{
		iter:        iter,
		hashNext:    iter.hasNext(),
		nextElement: iter.next(),
	}
}

func (it *PeekingIterator) hasNext() bool {
	return it.hashNext
}

func (it *PeekingIterator) next() int {
	ret := it.nextElement
	it.hashNext = it.iter.hasNext()
	if it.hashNext {
		it.nextElement = it.iter.next()
	}
	return ret
}

func (it *PeekingIterator) peek() int {
	return it.nextElement
}
